frost-core/src/frost/identifier.rs

```diff
@@ -14,183 +14,91 @@ use crate::{Ciphersuite, Error, Field, Group, Scalar};
 /// over, corresponding to some x-coordinate for a polynomial f(x) = y.  MUST NOT be zero in the
 /// field, as f(0) = the shared secret.
 #[derive(Copy, Clone)]
-pub struct Identifier<C: Ciphersuite>(pub(crate) Scalar<C>);
```

Comment:
OH. I see. Hm.

Reply:
We may want an NonZeroScalar style Scalar, which would just be Identifier, that wraps Scalar, and never can be zero or greater than the space of u16, but operates like a Scalar otherwise, so you can just use the Identifier type in artihmetic or whatever without converting it to a Scalar

https://github.com/zcash/orchard/blob/3faab98e9e82618a0f2d887054e9e28b0f7947dd/src/spec.rs#L95

Reply:
The hard part of doing that it's that there is no clean ciphersuite-generic way to convert a Scalar back to a uint16 (which is needed when serializing the identifier).

Using Identifier as a Scalar I think would require the `Deref` trait and the Rust docs is somewhat adamant about using it only for smart pointers. (See https://doc.rust-lang.org/std/ops/trait.Deref.html and also https://rust-lang.github.io/api-guidelines/predictability.html#only-smart-pointers-implement-deref-and-derefmut-c-deref. But some people disagree though, [see discussion](https://stackoverflow.com/questions/45086595/is-it-considered-a-bad-practice-to-implement-deref-for-newtypes) :rofl:  )

Reply:
> there is no clean ciphersuite-generic way to convert a Scalar back to a uint16 

True but the `u16` is an implementation detail, not a specification detail, and the spec does enforce `(De)serializeScalar`, which the `identifier` is in reality; so I think we get away with (`Try`)`From`'ing `Identifier` to `u16` by treating it as a `Scalar` that we enforce never has a value greater than `u16::MAX`; I'm going to try this out

Reply:
> True but the `u16` is an implementation detail, not a specification detail, and the spec does enforce `(De)serializeScalar`, which the `identifier` is in reality; so I think we get away with (`Try`)`From`'ing `Identifier` to `u16` by treating it as a `Scalar` that we enforce never has a value greater than `u16::MAX`; I'm going to try this out

I'm not sure about this... the spec states

> An identifier is an integer value associated with a participant, or signer, and is a value in the range [1, MAX_SIGNERS].

which is effectively a u16. Version 8 also uses `Scalar(identifier)` to convert an identifier to a scalar which wouldn't make sense if identifier was already a scalar.

Reply:
'not a specification detail' is not exactly correct but our rendering of Identifier is not fully pinned to the spec, if that makes sense

Reply:
https://github.com/ZcashFoundation/frost/pull/114

Reply:
> > True but the `u16` is an implementation detail, not a specification detail, and the spec does enforce `(De)serializeScalar`, which the `identifier` is in reality; so I think we get away with (`Try`)`From`'ing `Identifier` to `u16` by treating it as a `Scalar` that we enforce never has a value greater than `u16::MAX`; I'm going to try this out
> 
> I'm not sure about this... the spec states
> 
> > An identifier is an integer value associated with a participant, or signer, and is a value in the range [1, MAX_SIGNERS].
> 
> which is effectively a u16. Version 8 also uses `Scalar(identifier)` to convert an identifier to a scalar which wouldn't make sense if identifier was already a scalar.

`Scalar(id)` is interesting! This might need to be a per-ciphersuite function!

Reply:
> `Scalar(id)` is interesting! This might need to be a per-ciphersuite function! This does not explicitly deal in endianness buuuuut

We can do that, and use the implementation in this PR as the default implementation (or not)